ambry-store/src/main/java/com/github/ambry/store/RemoteTokenTracker.java

```diff
@@ -69,7 +103,201 @@ void refreshPeerReplicaTokens() {
   /**
    * @return a snapshot of peer replica to token map.
    */
-  Map<String, FindToken> getPeerReplicaAndToken() {
+  Map<String, Pair<Long, FindToken>> getPeerReplicaAndToken() {
     return new HashMap<>(peerReplicaAndToken);
   }
+
+  /**
+   * @return the file path which persists the remote token.
+   */
+  public String getPersistFilePath() {
+    return localReplica.getReplicaPath() + File.separator + REMOTE_TOKEN_FILE_NAME;
+  }
+
+  /**
+   * Provide a way to explicitly persist the token besides the background persistor
+   */
+  public boolean persistToken() {
```

Comment:
Just clarifying questions
1. Were we persisting the token on the sender before this ?
2. Why are we persisting now ?

Reply:
We don't persist the remote token currently. Remote token was introduced in the below PR and tried to solve the tombstone problem. It helps with the tombstone problem.
https://github.com/linkedin/ambry/pull/1796

If we don't persist the remote token, as long as there is server deployment and there is one replica down, we lose the remote token for that down replica. Because of that, we cannot remove any tombstone any more.
But if we persist the token, we can continue to remove tombstone as long as the token of the down replica has passed that tombstone.

We have more room to improve it after we persist the token. For example, we know how long the remote replicas has been down, we can decide if we remove the remote replica from the replica pool.